### PR TITLE
Pixel-based modelling with SLITronomy

### PR DESCRIPTION
**Pull request by aymgal**
This PR replaces #139 , with a much simpler integration to lenstronomy.

The general idea now is: when a source light profile such as `'SLIT_STARLETS'` is chosen, the `ImageModel` class (and subsequently `ImageLinearFit`) switches to "pixel-based modelling" mode. It effectively replaces the WLS that usually solves for linear amplitudes by the external SLITronomy solver.

In SLITronomy, three setups are supported:
- pixel-based modelling of source only (no lens light nor point source should be set)
- pixel-based modelling of source and lens light (no point source should be set)
- pixel-based modelling of source and point-sources light (no lens light should be set)

Supported light profiles are `'SLIT_STARLETS'` and `'SLIT_STARLETS_GEN2'`. The correctness of chosen profiles is checked by the SLITronomy utility `lenstronomy.Util.class_util.create_solver_class()`, which returns the right solver class (or raises errors if the configuration is not supported).

Differential extinction is not supported yet.

Additionally, you can add to the `kwargs_special` two sampled parameters to marginalise over offsets between the pixelated source grid and data grid.

I left the (drafted) old PR opened, that can help for the discussion (the number of commits is rather large, sorry, but is mainly due to the older PR...).

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -17,4 +17,7 @@ scikit-image
 #schwimmbad==0.3.0
 -e git://github.com/adrn/schwimmbad.git@master#egg=schwimmbad
 multiprocess>=0.70.8
-#-e git://github.com/sibirrer/fastell4py.git#egg=fastell4py
\ No newline at end of file
+#-e git://github.com/sibirrer/fastell4py.git#egg=fastell4py
+
+# switch to PyPi version once PR is merged
+-e git://github.com/aymgal/slitronomy.git@master#egg=slitronomy
```
Comment (sibirrer) [inline requirements.txt:23]: yes, I would also like to look into slitronomy and it's testing to make sure there is a stable distribution to comply with astropy affiliation

File: test/test_ImSim/test_Numerics/test_convolution.py [hunk 2]
```diff
@@ -26,6 +26,22 @@ def test_convolve2d(self):
         image_convolved = pixel_conv.convolution2d(self.model)
         npt.assert_almost_equal(np.sum(image_convolved), np.sum(self.model), decimal=2)
 
+    def test_copy_transpose(self):
```
Comment (sibirrer) [inline test/test_ImSim/test_Numerics/test_convolution.py:29]: the test coverage dropped about 2%. I don't comment on the tests but would be great to have good test coverage. There might be a few lines that can not be tested due to dependencies but most should be able to be tested (incl. raise statements).

**Comment by sibirrer** [comment]
travis failed after the update

**Comment by aymgal** [comment]
> travis failed after the update

Yes, I am not done with the changes (should not have pushed these commits actually 😉).

**Comment by sibirrer** [comment]
> > travis failed after the update
> 
> Yes, I am not done with the changes (should not have pushed these commits actually 😉).

no worries! just ping me once you think it's ready for me to review. Thanks :)

**Comment by aymgal** [comment]
The remaining big chunk of untested code are routines related to the external package pySAP (https://github.com/CEA-COSMIC/pysap) to perform efficient wavelet transforms.
When this package is not installed (or `force_no_pysap` set to `True`), python-only routines are called, which are tested.
In SLITronomy test routines, pySAP is used, so there it is properly tested. If you do want this to be tested also in lenstronomy as well, it would require changing the requirements/travis.yml to some extent to make sure pySAP is correctly compiled on the CI server (which sometimes can be really painful I must say...).

Otherwise please tell me anything needs to be improved :)

**Review by sibirrer — CHANGES_REQUESTED** [review]
Thanks a lot for the updates! Very close to merging. I have few clarification questions (see comments in code) and for the test coverage improvements. Looking at coveralls, is there a way to test the private functions _transform(), _inverse_transform() _coeffs2pysap()? I see that pySAP should not be imported for the tests and that this does not allow to test certain lines of code.
Then, once I approve the PR, before merging, can you release a new slitronomy PyPi version and update the PR with it such that I can merge with the sealed PyPi version.

File: lenstronomy/ImSim/image_linear_solve.py [hunk 3]
```diff
@@ -59,21 +70,51 @@ def _image_linear_solve(self, kwargs_lens=None, kwargs_source=None, kwargs_lens_
...
+            model, model_error, cov_param, param = self.image_pixelbased_solve(kwargs_lens, kwargs_source, 
+                                                                               kwargs_lens_light, kwargs_ps, 
+                                                                               kwargs_extinction, kwargs_special)
+        else:
+            A = self._linear_response_matrix(kwargs_lens, kwargs_source, kwargs_lens_light, kwargs_ps, kwargs_extinction, kwargs_special)
+            C_D_response, model_error = self._error_response(kwargs_lens, kwargs_ps, kwargs_special=kwargs_special)
+            d = self.data_response
+            param, cov_param, wls_model = de_lens.get_param_WLS(A.T, 1 / C_D_response, d, inv_bool=inv_bool)
+            _, _, _, _ = self.update_linear_kwargs(param, kwargs_lens, kwargs_source, kwargs_lens_light, kwargs_ps)
+            model = self.array_masked2image(wls_model)
```
[the remark this answers is not in the document]
Comment (aymgal) [reply lenstronomy/ImSim/image_linear_solve.py:94]: The model returned by the PixelSolver is already defined as a 2d image with likelihood mask applied, so no need here.

File: lenstronomy/ImSim/image_linear_solve.py [hunk 4]
```diff
@@ -59,21 +70,51 @@ def _image_linear_solve(self, kwargs_lens=None, kwargs_source=None, kwargs_lens_
...
+        By default, the linear parameters are computed with a weighted linear least square optimization (i.e. flux normalization of the brightness profiles)
+        However in case of pixel-based modelling, pixel values are constrained by an external solver (e.g. SLITronomy).
 
         :param kwargs_lens: list of keyword arguments corresponding to the superposition of different lens profiles
         :param kwargs_source: list of keyword arguments corresponding to the superposition of different source light profiles
         :param kwargs_lens_light: list of keyword arguments corresponding to different lens light surface brightness profiles
         :param kwargs_ps: keyword arguments corresponding to "other" parameters, such as external shear and point source image positions
         :param inv_bool: if True, invert the full linear solver Matrix Ax = y for the purpose of the covariance matrix.
+        This has no impact in case of pixel-based modelling.
         :return: 1d array of surface brightness pixels of the optimal solution of the linear parameters to match the data
```
[the remark this answers is not in the document]
Comment (aymgal) [reply lenstronomy/ImSim/image_linear_solve.py:82]: Indeed it's 2d, will fix that!

File: lenstronomy/ImSim/image_linear_solve.py [hunk 5]
```diff
@@ -59,21 +70,51 @@ def _image_linear_solve(self, kwargs_lens=None, kwargs_source=None, kwargs_lens_
...
+        computes the image (lens and source surface brightness with a given lens model) using the pixel-based solver.
+
+        :param kwargs_lens: list of keyword arguments corresponding to the superposition of different lens profiles
+        :param kwargs_source: list of keyword arguments corresponding to the superposition of different source light profiles
+        :param kwargs_lens_light: list of keyword arguments corresponding to different lens light surface brightness profiles
+        :param kwargs_ps: keyword arguments corresponding to point sources
+        :param kwargs_extinction: keyword arguments corresponding to dust extinction
+        :param kwargs_special: keyword arguments corresponding to "special" parameters
+        :param init_lens_light_model: optional initial guess for the lens surface brightness
+        :return: 1d array of surface brightness pixels of the optimal solution of the linear parameters to match the data
```
[the remark this answers is not in the document]
Comment (aymgal) [reply lenstronomy/ImSim/image_linear_solve.py:109]: will be fixed in next commits as well

File: lenstronomy/ImSim/image_linear_solve.py [hunk 6]
```diff
@@ -59,21 +70,51 @@ def _image_linear_solve(self, kwargs_lens=None, kwargs_source=None, kwargs_lens_
...
+        """
+        _, model_error = self._error_response(kwargs_lens, kwargs_ps, kwargs_special=kwargs_special)
+        model, param, _ = self.PixelSolver.solve(kwargs_lens, kwargs_source, kwargs_lens_light=kwargs_lens_light,
+                                                 kwargs_ps=kwargs_ps, kwargs_special=kwargs_special,
+                                                 init_lens_light_model=init_lens_light_model)
+        cov_param = None
+        kwargs_source, kwargs_lens_light = self.update_pixel_kwargs(kwargs_source, kwargs_lens_light)
         _, _, _, _ = self.update_linear_kwargs(param, kwargs_lens, kwargs_source, kwargs_lens_light, kwargs_ps)
```
[the remark this answers is not in the document]
Comment (aymgal) [reply lenstronomy/ImSim/image_linear_solve.py:117]: `point_source_only` is deprecated indeed... And yes duplication can be avoided here, thanks for seeing this

File: lenstronomy/ImSim/image_model.py [hunk 7]
```diff
@@ -51,12 +55,22 @@ def __init__(self, data_class, psf_class, lens_model_class=None, source_model_cl
...
-        self.num_bands = 1
         self._kwargs_numerics = kwargs_numerics
         if extinction_class is None:
             extinction_class = DifferentialExtinction(optical_depth_model=[])
         self._extinction = extinction_class
+        if kwargs_pixelbased is None:
+            kwargs_pixelbased = {}
+        self._pixelbased_bool = self._detect_pixelbased_models()
+        if self._pixelbased_bool is True:
+            from slitronomy.Util.class_util import create_solver_class  # requirement on SLITronomy is exclusively here
```
[the remark this answers is not in the document]
Comment (aymgal) [reply lenstronomy/ImSim/image_model.py:66]: Yes I reduced it to this single import. And this way there no obligation to install SLITronomy as well when no using corresponding light profiles

File: lenstronomy/LightModel/Profiles/starlets.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        return coeffs_list
+
+    def _load_pysap(self, force_no_pysap):
+        """load pySAP module"""
+        if force_no_pysap:
+            return False, None
+        try:
+            import pysap
+        except ImportError:
+            return False, None
```
[the remark this answers is not in the document]
Comment (aymgal) [reply lenstronomy/LightModel/Profiles/starlets.py:165]: Here is a quick notebook timeit, for decomposing a 100x100 image into starlet coefficients:
__with pySAP__: 452 µs ± 69.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
__with pure python__: 1.47 ms ± 44.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In SLITronomy starlet functions are called _a lot_ of times, so it plays significant role.
When only using the `SLIT_Starlets` profile for a few calls of decomposition/reconstruction methods, then it does not matter (and starts to matter again for really big images).

[the remark this answers is not in the document]
Comment (aymgal) [reply lenstronomy/LightModel/Profiles/starlets.py:165]: Indeed a warning can be much more explicit, I will propose something in that direction

**Comment by aymgal** [comment]
I further completed test coverage.
Concerning `_transform` and `_inverse_transform`... I am not sure how those can be tested, as they almost exclusively call pysap-related methods (now that `pysap2coeffs`/`coeffs2pysap` are tested). Is there some clever technique to deal with this?

**Review by sibirrer — APPROVED** [review]
This looks very good and I can formally approve it. The tests are sufficient to the level that they are possible. Thanks for putting your effort into this!
So now there are two remaining minor things:
- Warning if pySAP is not installed
- pointer to PyPi version of SLITronomy
I also looked at SLITronomy and it looks really good with good test coverage and documentation. One thing thought might be changed when releasing it on PyPi is the status (from beta to stable), if you feel comfortable. I know it's primarily beta because it still requires your fork of lenstronomy to run. The astropy community would probably like to see that dependencies are marked as stable. So I formally approve this PR and wait for the merge for these minor changes. Looking very much forward to push the merge button!

**Comment by aymgal** [comment]
If added the warning message and released a stable PyPi version 0.3.2.

File: requirements.txt [hunk 9]
```diff
@@ -18,6 +18,4 @@ scikit-image
 -e git://github.com/adrn/schwimmbad.git@master#egg=schwimmbad
 multiprocess>=0.70.8
 #-e git://github.com/sibirrer/fastell4py.git#egg=fastell4py
-
-# switch to PyPi version once PR is merged
--e git://github.com/aymgal/slitronomy.git@master#egg=slitronomy
+slitronomy
```
Comment (sibirrer) [inline requirements.txt:21]: Can you add a specific version number? This makes it more stable and makes sure that when a new slitronomy version is released, it goes through a PR to check compatibility.

**Review by sibirrer — CHANGES_REQUESTED** [review]

**Comment by aymgal** [comment]
I realize that the `image_pixelbased_solve` has to support stand-alone use as well (without calling `image_linear_solve`), typically for its argument `init_lens_light_model`. So I move back the call to `update_linear_kwargs` inside the method, despite the small duplication code... hope it's still ok like this

File: requirements.txt (hunk 9 above)
Comment (aymgal) [reply requirements.txt:21]: Ok I will add it

**Review by sibirrer — APPROVED** [review]

**Comment by sibirrer** [comment]
Very nice! I will merge now and then make a new lenstronomy PyPi release (overdue anyways).

**Pull request merged**